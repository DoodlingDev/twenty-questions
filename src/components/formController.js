// @flow
import React, { Component } from "react";
import FormNodeObject from "./nodes/types/object";
import getWidgets from "../utils/getWidgets";

type q20$FormControllerProps = {
  schema: q20$Schema,
  title: string,
  description?: string,
  widgets?: q20$RenderedNode[],
};

type q20$Schema = {
  title: string,
  description?: string,
  properties: q20$Node[],
  widget?: string,
};

type q20$FormControllerState = {
  values: q20$FormValues,
  errors: q20$FormErrors,
};

type q20$FormValues = {
  [key: string]: any,
};

type q20$FormErrors = {
  [key: string]: q20$Error,
};

/**
 * FormController
 *   master component for the q20 forms, holds values and errors
 */
export default class FormController extends Component<
  q20$FormControllerProps,
  q20$FormControllerState,
> {

  /**
   * state
   * @param {object} values input values from the form
   * @param {object} errors error objects created by the validator
   */
  state = {
    values: {},
    errors: {},
  };

  /**
   * constructor
   *
   * @param {FormControllerProps} props
   */
  constructor(props: q20$FormControllerProps) {
    super(props);
  }

  /**
   * render
   *   react render function
   *   renders the FormBuilder to make the markup
   * @return {React$Element} FormBuilder
   */
  render() {
    const { widgets } = this.props;
    const { title, description, properties, widget } = this.props.schema;
    return (
      <form>
        <h2>{title}</h2>
        {description && <h3>{description}</h3>}

        <FormNodeObject
          name={title}
          label={title}
          description={description}
          type={"object"}
          path={`${title}`}
          properties={properties}
          widgets={getWidgets(widgets)}
          widget={widget ? widget : undefined}
        />
      </form>
    );
  }
}
