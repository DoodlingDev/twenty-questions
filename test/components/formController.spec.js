/* eslint require-jsdoc: "off" */
import React from "react";
import { FormController } from "../../src/components/formController.js";
import renderer from "react-test-renderer";
import { /*  shallow,  */ mount } from "enzyme";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({ adapter: new Adapter() });

const setupProps = {
  title: "test form title",
  description: "this is a test form",
  validate: {
    single: jest.fn(),
    state: {},
  },
  properties: [
    {
      // this name is left with a space specifically to
      // test that names get camelized
      name: "test name",
      type: "object",
      properties: [
        {
          type: "string",
          name: "testString",
        },
      ],
    },
  ],
};

it("renders correctly", () => {
  const render = renderer.create(<FormController {...setupProps} />).toJSON();
  expect(render).toMatchSnapshot();
});

function setup(renderFn, testProps = setupProps) {
  return renderFn(<FormController {...testProps} />);
}

describe("changeValue", () => {
  it("should update when given the correct input", () => {
    const wrapper = setup(mount);
    wrapper.instance().changeValue({
      path: "test.path",
      name: "test_name",
      value: "test value",
    });
    wrapper.update();
    expect(wrapper.state().values).toMatchObject({
      "test.path": "test value",
    });
  });

  it("shouldn't change the state when incorrect input is passed", () => {
    const wrapper = setup(mount);
    wrapper.instance().changeValue({
      path: "test.path",
      name: "test_name",
    });
    wrapper.update();
    expect(wrapper.state().values).not.toMatchObject({
      "test.path": undefined,
    });
  });

  it("should call validate single when typeAhead is true", () => {
    const wrapper = setup(mount);
    wrapper.instance().props.validate.single = jest.fn();
    wrapper.instance().changeValue({
      path: "test.path",
      name: "test_name",
      value: "boop",
    });
    wrapper.update();
    expect(wrapper.instance().props.validate.single).toHaveBeenCalledWith({
      path: "test.path",
      name: "test_name",
      value: "boop",
    });
  });

  it("should not call validate single when typeAhead is false", () => {
    const wrapper = setup(mount, { typeAheadValidation: false, ...setupProps });
    wrapper.instance().props.validate.single = jest.fn();
    wrapper.instance().changeValue({
      path: "test.path",
      name: "test_name",
      value: "boop",
    });
    wrapper.update();
    expect(wrapper.instance().props.validate.single).not.toHaveBeenCalled();
  });
});

describe("registerField", () => {
  it("should register each field when they are rendered", () => {
    const wrapper = setup(mount);
    wrapper.update();
    expect(wrapper.state().fieldRegistry.length).toEqual(2);
    expect(wrapper.state().fieldRegistry).toContain(
      "testFormTitle.testName.testString",
    );
  });

  it("should update with unique paths", () => {
    const wrapper = setup(mount);
    const n = wrapper.state().fieldRegistry.length;
    wrapper.instance().registerField("purple.people.eater");
    wrapper.update();
    expect(wrapper.state().fieldRegistry.length).toEqual(n + 1);
  });

  it("should not update with non-unique paths", () => {
    const wrapper = setup(mount);
    const n = wrapper.state().fieldRegistry.length;
    wrapper.instance().registerField(wrapper.state().fieldRegistry[0]);
    wrapper.update();
    expect(wrapper.state().fieldRegistry.length).toEqual(n);
  });
});

describe("deleteRowValue", () => {
  const arrayValue = {
    "testFormTitle.testName.testArray": {},
    "testFormTitle.testName.testArray.0.a": {},
    "testFormTitle.testName.testArray.0.a.one": "A",
    "testFormTitle.testName.testArray.0.a.two": "B",
    "testFormTitle.testName.testArray.1.a": {},
    "testFormTitle.testName.testArray.1.a.one": "AA",
    "testFormTitle.testName.testArray.1.a.two": "BB",
    "testFormTitle.testName.testArray.2.a": {},
    "testFormTitle.testName.testArray.2.a.one": "AAA",
    "testFormTitle.testName.testArray.2.a.two": "BBB",
    "testFormTitle.testName.testArray.3.a": {},
    "testFormTitle.testName.testArray.3.a.one": "AAAA",
    "testFormTitle.testName.testArray.3.a.two": "BBBB",
  };

  it("should remove the row in question", () => {
    const wrapper = setup(mount);
    wrapper.instance().state.values = { ...arrayValue };
    const valueLength = Object.keys(wrapper.state().values).length;
    wrapper.instance().deleteRowValue({
      path: "testFormTitle.testName.testArray",
      index: 1,
    });
    wrapper.update();
    expect(Object.keys(wrapper.state().values).length).toBe(valueLength - 3);
  });

  it("should preserve data", () => {
    const wrapper = setup(mount);
    wrapper.instance().state.values = { ...arrayValue };
    wrapper.instance().deleteRowValue({
      path: "testFormTitle.testName.testArray",
      index: 1,
    });
    wrapper.update();
    const allCorrect = (function() {
      const values = wrapper.state().values;

      for (let i = 0, l = 3; i < l; i++) {
        if (
          values[`testFormTitle.testName.testArray.${i}.a.one`].length !==
          values[`testFormTitle.testName.testArray.${i}.a.two`].length
        ) return false;

        if (
          !/A+/.test(values[`testFormTitle.testName.testArray.${i}.a.one`]) ||
          !/B+/.test(values[`testFormTitle.testName.testArray.${i}.a.two`])
        ) return false;
      }
      return true;
    })();
    expect(allCorrect).toBeTruthy();
  });

  it("should decrement the numbers properly", () => {
    const wrapper = setup(mount);
    wrapper.instance().state.values = { ...arrayValue };
    wrapper.instance().deleteRowValue({
      path: "testFormTitle.testName.testArray",
      index: 1,
    });
    wrapper.update();
    const hasNoThrees = (function(){
      for (let keyName in wrapper.state().values) {
        if (/3/.test(keyName)) return false;
      }
      return true;
    })();
    expect(hasNoThrees).toBeTruthy();
  });
});

