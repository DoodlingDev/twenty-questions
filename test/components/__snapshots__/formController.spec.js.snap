// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form>
  <h2>
    test form title
  </h2>
  <h3>
    this is a test form
  </h3>
  <div>
    <span>
      this is a test form
    </span>
    <div>
      <div
        className="q20_node__string"
      >
        <div
          className="q20_testString__errorHandler"
        >
          <input
            className="q20_nodeString__testString--input"
            id="testFormTitle.testName.testString"
            onChange={[Function]}
            placeholder={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</form>
`;
